core/cache/application_test.go

```diff
@@ -21,6 +22,65 @@ func (s *ApplicationSuite) SetUpTest(c *gc.C) {
 	s.entitySuite.SetUpTest(c)
 }
 
+func (s *ApplicationSuite) TestFieldWatcherStops(c *gc.C) {
+	m := s.newApplication(appChange)
+	w := m.WatchFields()
+	wc := NewNotifyWatcherC(c, w)
+	// Sends initial event.
+	wc.AssertOneChange()
+	wc.AssertStops()
+}
+
+func (s *ApplicationSuite) TestFieldWatcherAnyChange(c *gc.C) {
+	m := s.newApplication(appChange)
+	w := m.WatchFields()
+	defer workertest.CleanKill(c, w)
+
+	wc := NewNotifyWatcherC(c, w)
+	// Sends initial event.
+	wc.AssertOneChange()
+
+	change := appChange
+	change.Config = map[string]interface{}{
+		"key": "changed",
+	}
+	m.SetDetails(change)
+	wc.AssertOneChange()
+}
+
+func (s *ApplicationSuite) TestFieldWatcherSpecificChange(c *gc.C) {
+	m := s.newApplication(appChange)
+	w := m.WatchFields(cache.ApplicationCharmURLChanged, cache.ApplicationExposedChanged)
+	defer workertest.CleanKill(c, w)
+
+	wc := NewNotifyWatcherC(c, w)
+	// Sends initial event.
+	wc.AssertOneChange()
+
+	change := appChange
+
+	// Not a change we are watching for.
+	change.Config = map[string]interface{}{
+		"key": "changed",
+	}
+	m.SetDetails(change)
+	wc.AssertNoChange()
+
+	change.CharmURL = "www.new-charm-url.com"
```

Comment:
often we use something like .invalid in test data, just in case it escapes somewhere, we don't end up with anything like a valid URL.

Reply:
or example.com domain